docs/topics/signals.rst

```diff
@@ -158,6 +158,23 @@ item_error
     :param failure: the exception raised
     :type failure: twisted.python.failure.Failure
 
+bytes_received
+--------------
+
+.. signal:: bytes_received
+.. function:: bytes_received(data, request)
+
+    Sent by the HTTP 1.1 download handler when a group of bytes is
```

Comment:
I wonder if a signal should be renamed, if it is just for HTTP 1.1 download handler. 

Is it sent e.g. for S3 responses, or for FTP responses? It could be, because S3DownloadHandler uses http download handler,though I haven't checked it. If it is sent, we should document that. If it is not sent for all handlers, imagine in future we decide to add support for them (e.g. S3 handler starts to send this signal). It'd be a backwards incompatible change, as user code may rely on bytes_received to be only issues by HTTP 1.1 download handler, not by S3 or FTP. Having different signal names can avoid this confusion. Alternatively, we can decide to support everything from the beginning.

Reply:
Just pushed 8beb77c3f33eeb16b42719fa3a263c840ddaaf49, which adds a `source` parameter to the signal (current values could be "http11" or "s3"). I'm not sure how to test the occurrence of the "s3" value without actually requesting stuff from S3.
Let me know what you think of this approach @kmike.

Reply:
Let's check pro/cons of these aproaches, from user point of view.

**Use case 1**: user wants to check traffic received from websites. Two signals:

```py
def on_bytes_received_http(data):
    # ...

crawler.signals.connect(on_bytes_received_http, signals.bytes_received_http)
```

One signal, "source" field:

```py
def on_bytes_received(data, source):
    if source != 'http11':
        return
    # ...

crawler.signals.connect(on_bytes_received, signals.bytes_received)
```

**Use case 2**: user wants to check traffic received from websites and from downloading s3 files (e.g. with seed urls used in start_requests?), but wants to handle them differently:

```py
def on_bytes_received_http(data):
    # ...

def on_bytes_received_s3(data):
    # ...

crawler.signals.connect(on_bytes_received_http, signals.bytes_received_http)
crawler.signals.connect(on_bytes_received_s3, signals.bytes_received_s3)
```

One signal, "source" field:

```py
def on_bytes_received(data, source):
    if source == 'http11':
        # ...
    elif source == 's3':
        # ...

crawler.signals.connect(on_bytes_received, signals.bytes_received)
```

**Use case 3**: user wants overall downloaded data stats, source doesn't matter.

Two signals:

```py
def on_bytes_received(data):
    # ...

crawler.signals.connect(on_bytes_received, signals.bytes_received_http)
crawler.signals.connect(on_bytes_received, signals.bytes_received_s3)
```

One signal, "source" field:

```py
def on_bytes_received(data, source):
    # ...

crawler.signals.connect(on_bytes_received, signals.bytes_received)
```

To my taste, "two signals" approach 

* wins in Use case 1 (shorter, less error prone)
* wins in Use case 2 (less error prone, but more code)
* loses in Use case 3 (more code, need to find all signals related to downloading)

In first two use cases "less error prone" means that if you make a typo in a signal name, an error will be raised (there is no such a signal), while if you check against non-existing source, it will silently work wrong. Because of this "less error prone" I think having 2 separate signals is better than relying on "source" field.

For the stats use case a single signal could make sense, which is fired if any other bytes_received signal are fired. It will also allow to add more signal sources without users having to update their code. It can actually be the current bytes_received signal. But I wonder what's the performance overhead of that, having to send 2x signals on every chunk, as compared to benefit (it may still worth it).

---

One more thing, I'm not sure http11 is a good name here, because it may be set for http1.0 traffic as well - though I may be wrong :) I'm assuming that HTTP11DownloadHandler handles communication with all kind of of http servers, but maybe I'm wrong.

Reply:
I was in favor of multiple signals, but re-reading I’m starting to think that a single signal is much cleaner and should work, as long as we clearly document that additional download handlers may support this signal in the future, and recommend to use the `request` object to filter out non-HTTP traffic if desired.

This does not allow to know whether the traffic comes from HTTP 1.0, 1.1 or 2.0. Do we want to allow that? I don’t think we need that at the moment, and in the future we could implement a separate, new signal, e.g. `protocol_negotiated` or similar, that provides this information along with the affected request.

Reply:
That's absolutely true, thanks for pointing it out. The request object should be enough to determine whether or not the source is S3.
Regarding the version, we do not support HTTP/2 as far as I remember, and HTTP/1.0 needs to be explicitly enabled by changing the download handler. +1 on the `protocol_negotiated` signal once we add support for HTTP/2.
I'll go ahead and simplify the implementation, thanks both ❤️

Reply:
@kmike These would be the 3 use cases with a single signal and relying on `request.url`:

**Use case 1**: user wants to check traffic received from websites:

```py
def on_bytes_received(data, request, spider):
    if urlparse(request.url).protocol not in {'http', 'https'}:
        return
    # …

crawler.signals.connect(on_bytes_received, signals.bytes_received)
```

**Use case 2**: user wants to check traffic received from websites and from downloading s3 files (e.g. with seed urls used in start_requests?), but wants to handle them differently:

```py
def on_bytes_received(data, request, spider):
    protocol = urlparse(request.url).protocol
    if protocol in {'http', 'https'}:
        # …
    elif protocol == 's3':
        # …

crawler.signals.connect(on_bytes_received, signals.bytes_received)
```

**Use case 3**: user wants overall downloaded data stats, source doesn't matter:

```py
def on_bytes_received(data, request, spider):
    # …

crawler.signals.connect(on_bytes_received, signals.bytes_received)
```

Reply:
Ok, fair enough. Let's proceed with a single signal, without source :)

Reply:
Re examples, nitpick (in case we decide to put them somewhere, or if someone gets to this page via search) - they should use `scheme = urlparse_cached(request).scheme`, like scrapy's DownloadHandlers.